pkg/api/defaults-custom-cloud-profile.go

```diff
@@ -93,100 +110,114 @@ func (cs *ContainerService) SetCustomCloudProfileEnvironment() error {
 			}
 
 			env.ManagementPortalURL = endpoints.PortalEndpoint
-			env.ResourceManagerVMDNSSuffix = fmt.Sprintf("cloudapp.%s", azsFQDNSuffix)
-			env.StorageEndpointSuffix = fmt.Sprintf("%s.%s", cs.Location, azsFQDNSuffix)
-			env.KeyVaultDNSSuffix = fmt.Sprintf("vault.%s.%s", cs.Location, azsFQDNSuffix)
+
+			if (p.IsAzureStackCloud()) {
+				azsFQDNSuffix := getAzureStackFQDNSuffix(p.CustomCloudProfile.PortalURL, cs.Location)
+				env.ResourceManagerVMDNSSuffix = fmt.Sprintf("cloudapp.%s", azsFQDNSuffix)
+				env.StorageEndpointSuffix = fmt.Sprintf("%s.%s", cs.Location, azsFQDNSuffix)
+				env.KeyVaultDNSSuffix = fmt.Sprintf("vault.%s.%s", cs.Location, azsFQDNSuffix)
+			} else if (env.ResourceManagerVMDNSSuffix == "" || env.StorageEndpointSuffix == "" || env.KeyVaultDNSSuffix == "") {
+				// Non-AzureStack CustomCloud MUST provide suffixes			
+				return fmt.Errorf("Non-AzureStack CustomCloudProfile MUST provide ResourceManagerVMDNSSuffix, StorageEndpointSuffix, KeyVaultDNSSuffix")
+			}
 		}
 	}
+
 	return nil
 }
 
-// AzureStackCloudSpecParams is the parameters when we set the azure stack cloud spec defaults for ContainerService.
-type AzureStackCloudSpecParams struct {
+// AzureCustomCloudSpecParams is the parameters when we set the azure custom cloud spec defaults for ContainerService.
+type AzureCustomCloudSpecParams struct {
 	IsUpgrade bool
 	IsScale   bool
 }
 
-// SetAzureStackCloudSpec sets the cloud spec for Azure Stack .
-func (p *Properties) SetAzureStackCloudSpec(params AzureStackCloudSpecParams) error {
-	if p.IsAzureStackCloud() {
-		var azureStackCloudSpec AzureEnvironmentSpecConfig
+// SetAzureCustomCloudSpec sets the cloud spec for Custom Cloud .
+func (p *Properties) SetCustomCloudSpec(params AzureCustomCloudSpecParams) error {
+	if p.IsCustomCloudProfile() {
+		var azureCustomCloudSpec AzureEnvironmentSpecConfig
 		switch p.CustomCloudProfile.DependenciesLocation {
```

Comment:
You would need predefine a AzureEnvironmentSpecConfig for the new security cloud and set the azureCustomCloudSpec based on the location.

Reply:
This is a custom cloud and location can literally be anything. I do not want to (probably not even allowed) add specific new cloud-related information in the public repo. Instead, everything should come from user-provided config...

Reply:
The following is the AzureChinaCloudSpec for china cloud. as you can see, all the dependencies including to images and plugins are hosted separately in Azure China cloud.
end customer will not know the values.


//AzureChinaCloudSpec is the configurations for Azure China (Mooncake)
	AzureChinaCloudSpec = AzureEnvironmentSpecConfig{
		CloudName: AzureChinaCloud,
		//DockerSpecConfig specify the docker engine download repo
		DockerSpecConfig: DockerSpecConfig{
			DockerEngineRepo:         "https://mirror.azk8s.cn/docker-engine/apt/repo/",
			DockerComposeDownloadURL: "https://mirror.azk8s.cn/docker-toolbox/linux/compose",
		},
		//KubernetesSpecConfig - Due to Chinese firewall issue, the default containers from google is blocked, use the Chinese local mirror instead
		KubernetesSpecConfig: KubernetesSpecConfig{
			KubernetesImageBase:              "gcr.azk8s.cn/google_containers/",
			TillerImageBase:                  "gcr.azk8s.cn/kubernetes-helm/",
			ACIConnectorImageBase:            "dockerhub.azk8s.cn/microsoft/",
			NVIDIAImageBase:                  "dockerhub.azk8s.cn/nvidia/",
			AzureCNIImageBase:                "dockerhub.azk8s.cn/containernetworking/",
			MCRKubernetesImageBase:           "mcr.microsoft.com/k8s/core/",
			CalicoImageBase:                  "dockerhub.azk8s.cn/calico/",
			EtcdDownloadURLBase:              "https://mirror.azk8s.cn/kubernetes/etcd",
			KubeBinariesSASURLBase:           DefaultKubernetesSpecConfig.KubeBinariesSASURLBase,
			WindowsTelemetryGUID:             DefaultKubernetesSpecConfig.WindowsTelemetryGUID,
			CNIPluginsDownloadURL:            "https://mirror.azk8s.cn/kubernetes/containernetworking-plugins/cni-plugins-amd64-" + CNIPluginVer + ".tgz",
			VnetCNILinuxPluginsDownloadURL:   "https://mirror.azk8s.cn/kubernetes/azure-container-networking/azure-vnet-cni-linux-amd64-" + AzureCniPluginVerLinux + ".tgz",
			VnetCNIWindowsPluginsDownloadURL: "https://mirror.azk8s.cn/kubernetes/azure-container-networking/azure-vnet-cni-windows-amd64-" + AzureCniPluginVerWindows + ".zip",
			ContainerdDownloadURLBase:        "https://mirror.azk8s.cn/kubernetes/containerd/",
		},
		DCOSSpecConfig: DCOSSpecConfig{
			DCOS188BootstrapDownloadURL:     fmt.Sprintf(AzureChinaCloudDCOSBootstrapDownloadURL, "5df43052907c021eeb5de145419a3da1898c58a5"),
			DCOSWindowsBootstrapDownloadURL: "https://dcosdevstorage.blob.core.windows.net/dcos-windows",
			DCOS190BootstrapDownloadURL:     fmt.Sprintf(AzureChinaCloudDCOSBootstrapDownloadURL, "58fd0833ce81b6244fc73bf65b5deb43217b0bd7"),
			DCOS198BootstrapDownloadURL:     fmt.Sprintf(AzureChinaCloudDCOSBootstrapDownloadURL, "f4ae0d20665fc68ee25282d6f78681b2773c6e10"),
		},

		EndpointConfig: AzureEndpointConfig{
			ResourceManagerVMDNSSuffix: "cloudapp.chinacloudapi.cn",
		},
		OSImageConfig: map[Distro]AzureOSImageConfig{
			Ubuntu:            Ubuntu1604OSImageConfig,
			Ubuntu1804:        Ubuntu1804OSImageConfig,
			RHEL:              RHELOSImageConfig,
			CoreOS:            CoreOSImageConfig,
			AKSUbuntu1604:     AKSUbuntu1604OSImageConfig,
			AKS1604Deprecated: AKSUbuntu1604OSImageConfig, // for back-compat
			AKSUbuntu1804:     AKSUbuntu1804OSImageConfig,
			AKS1804Deprecated: AKSUbuntu1804OSImageConfig, // for back-compat
		},
	}

Reply:
I do not think it is allowed to put LX-specific URLs in anywhere in public clouds. I also believe all dependency plugins are pre-installed in aks-engine base images so that disconnected clouds will need to go fetch them, that is why I think it is fine to use AzurePublicCloud defaults for custom clouds for now.

Reply:
I am not aware of the requirement that the LX endpoints are also confidential.
and if we will rely on the VHD to include all dependencies. I am ok with current change.